### PR TITLE
Add testing of FileStream not truncating files when opening fails

### PR DESCRIPTION
**Pull request by poizan42**
Test of #24081 (succeeds after dotnet/coreclr#14043)

### CODE REVIEW

**Comment by poizan42** [comment]
Is it overkill to have the test being run by all ctors with FileShare?

**Comment by karelz** [comment]
@poizan42 thanks for your contribution!
Can you please update the title to describe the change and add description in top post with issue links? (issue numbers in titles are rarely useful - think about someone in future looking at history ;)). Thanks!

I have added you as collaborator, so that we can assign the issue #24081 to you (I assume this is a fix for it in combo with https://github.com/dotnet/coreclr/pull/14043). It is GH limitation for assignments :(.
Please note that it will auto-subscribe you to all repos notifcactions, which is madness of ~500 notifications per day. We recommend everyone to disable them and watch only notifications for your mentions.

**Comment by poizan42** [comment]
@stephentoub I have fixed those now

File: src/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs [hunk 1]
```diff
@@ -117,5 +118,26 @@ public void FileShareOpenOrCreate()
...
+        [InlineData(FileMode.Truncate)]
+        public void NoTruncateOnFileShareViolation(FileMode fileMode)
+        {
+            string fileName = GetTestFilePath();
+
+            using (FileStream fs = CreateFileStream(fileName, FileMode.Create, FileAccess.ReadWrite, FileShare.None))
+            {
+                fs.Write(new byte[] { 42 }, 0, 1);
+                fs.Flush();
+                FSAssert.ThrowsSharingViolation(() => CreateFileStream(fileName, fileMode, FileAccess.Write, FileShare.None).Dispose());
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs]: Nit: the Dispose shouldn't be necessary

Comment (poizan42) [reply src/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs]: But if it doesn't throw then we leave the file open until a GC occurs, does that not matter?

Comment (stephentoub) [reply src/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs]: > But if it doesn't throw

That's true, but if it doesn't throw, we have bigger problems :)  I'm ok with it either way, but in general we haven't worried about such things elsewhere.

**Comment by pjanotti** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Pull request merged**